contracts/settling_game/modules/crypts/CryptRun.cairo

```diff
@@ -0,0 +1,277 @@
+
+@storage_var
+func crypt(token_id: Uint256) -> (crypt: Crypt) {
+}
+
+@storage_var
+func adventurer(token_id: Uint256) -> (location: AdventurerLocation) {
+}
+
+// -----------------------------------
+// INITIALIZER & UPGRADE
+// -----------------------------------
+
+@external
+func initializer{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    address_of_controller: felt, proxy_admin: felt
+) {
+    Module.initializer(address_of_controller);
+    Proxy.initializer(proxy_admin);
+    return ();
+}
+
+@external
+func upgrade{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    new_implementation: felt
+) {
+    Proxy.assert_only_admin();
+    Proxy._set_implementation_hash(new_implementation);
+    return ();
+}
+
+// -----------------------------------
+// EXTERNAL
+// -----------------------------------
+
+// @notice Opens crypt.... Checks if the Crypt is active. If < 24hrs or > 10 successful quests, function creates a new seed.
+// @param crypt_id: Crypts erc721 ID
+// @param adventurer_id: Adventurer erc721 ID
+// @returns seed: Random seed. This is used in the Client to reconstruct the Crypt in a visual form.
+@external
+func open_crypt{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    crypt_id: Uint256, adventurer_id: Uint256
+) -> (seed: felt) {
+    // check ownership
+    assert_adventurer(adventurer_id);
+
+    let (now) = get_block_timestamp();
+
+    let (crypt_status) = crypt.read(crypt_id);
+    let is_seed_active = is_le(now, crypt_status.time_stamp);
+
+    if (is_seed_active == FALSE) {
+        // get random
+        tempvar randomSeed = 123456842657854254785;
+
+        // write crypt set ts 1 day from now
+        crypt.write(crypt_id, Crypt(now + DAY, randomSeed, NUMBER_SUCCESSFUL_RUNS));
+
+        // write adventurer
+        adventurer.write(
+            adventurer_id, AdventurerLocation(crypt_id.low, NODE_START_INDEX, randomSeed)
+        );
+
+        return (seed=randomSeed);
+    }
+
+    // we save the seed within the adventurer struct, this is incase the seed changes midrun the player can still complete
+    adventurer.write(
+        adventurer_id, AdventurerLocation(crypt_id.low, NODE_START_INDEX, crypt_status.seed)
+    );
+
+    return (seed=crypt_status.seed);
+}
+
+// @notice Move Adventurer within the Crypt.
+// @param crypt_id: Crypts erc721 ID
+// @param adventurer_id: Adventurer erc721 ID
+// @param start_location: Index of the current location
+// @param end_location: Index of the location wishing to travel to
+@external
+func move_to_location_in_crypt{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    crypt_id: Uint256, adventurer_id: Uint256, start_location: felt, end_location: felt
+) -> () {
+    alloc_locals;
+    // check ownership
+    assert_adventurer(adventurer_id);
+
+    // assert at location requested
+    let (adventurer_location) = assert_and_return_location(crypt_id, adventurer_id, start_location);
```

Comment:
Since adventurer's location is stored, why assert it? Why even pass the `start_location`? The adventurer can't be at any other place / vertex no?

Reply:
true. Will refactor.